### PR TITLE
Set time plugin

### PR DESCRIPTION
**Pull request by Somac**
Oskari.app.playBundle(
{
  bundlename : 'shadow-plugin-3d'
});

Shadowing plugin for 3d maps. Gets time and date from user and changes shadowing on map accordingly.

### CODE REVIEW

File: bundles/mapping/shadow-plugin-3d/instance.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return this.sandbox;
+        },
+        start: function (sandbox) {
+            if (this._started) {
+                return;
+            }
+            this._started = true;
+            sandbox = sandbox || Oskari.getSandbox();
+            this.setSandbox(sandbox);
+            this._mapmodule = sandbox.findRegisteredModuleInstance('MainMapModule');
```
Comment (ZakarFin) [inline bundles/mapping/shadow-plugin-3d/instance.js]: We could check if the map module supports 3D and stop if it doesn't

File: bundles/mapping/shadow-plugin-3d/view/ShadowTool/ShadowTool.jsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    return (
+        <Background>
+            <Row>
+                <Col>
+                    <StyledIcon type="calendar" style={{ color: '#d9d9d9', fontSize: '18px' }} />
+                    <StyledInput value={dateValue} onChange={changeDate} />
+                </Col>
+                <Col>
+                    <StyledButton onClick={setCurrentTime}>Nykyhetki</StyledButton>
```
Comment (Somac) [inline bundles/mapping/shadow-plugin-3d/view/ShadowTool/ShadowTool.jsx]: I will replace this hardcoded string on the next ticket. Already made so much changes to this file on the next part so too much hassle merging these branches if I make the change here.

File: bundles/mapping/shadow-plugin-3d/plugin/ShadowingPluginClass.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    _addToMobileToolBar () {
+        const resetMapStateControl = jQuery('.toolbar_mobileToolbar').find('.mobile-reset-map-state');
+        jQuery(this._element).insertAfter(resetMapStateControl);
+    }
+
+    _addToPluginContainer () {
+        const panButtonsControl = jQuery('.mappluginsContent').find('.panbuttons');
+        jQuery(this._element).insertAfter(panButtonsControl);
+    }
```
Comment (ZakarFin) [inline bundles/mapping/shadow-plugin-3d/plugin/ShadowingPluginClass.js]: Use "index" to assign "priority"/placement among other plugins: https://github.com/oskariorg/oskari-frontend/blob/master/bundles/framework/coordinatetool/plugin/CoordinateToolPlugin.js#L29

File: bundles/mapping/shadow-plugin-3d/view/ShadowTool/ShadowTool.jsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    const validateDate = (target) => {
+        const matches = /^(0[1-9]|1[0-9]|2[0-9]|3[0-1]|[1-9])[/](0[1-9]|1[0-2]|[1-9])$/.exec(target);
+        if (matches === null) {
+            return false;
+        }
+        const d = parseInt(matches[1]);
+        const m = matches[2] - 1;
+        const dateObject = new Date(2019, m, d);
+        return dateObject.getDate() === d && dateObject.getMonth() === m;
+    };
```
Comment (ZakarFin) [inline bundles/mapping/shadow-plugin-3d/view/ShadowTool/ShadowTool.jsx]: Add a comment that these are duplicated in the SetTimeRequestHandler. Perfect option would be to have some util.js on 3d-mapmodule that could be imported and used both here and in the requesthandler but the comment is enough for now.

Comment (ZakarFin) [reply bundles/mapping/shadow-plugin-3d/view/ShadowTool/ShadowTool.jsx]: There will be a follow up PR that has changes to this so merging for now.

**Pull request merged**